### DISCUSSION TITLE
Can not auto-import local files

### DISCUSSION THREAD

**Discussion by jaklimoff** [microsoft/pyright#7631] [Q&A] [answered]
Hello. I have a strange problem with my pyright. Currently im trying to make it work on LazyVim but i checked also on other vim editions and VSCode and keep getting same result. For some weird reason, pyright doesnt allow me to auto-import classes from local files, but works perfectly for builtin libraries and venv.

I think the problem is in config or maybe some weird cached settings, because it stopped working in every project i have.  I created a small draft project to showcase my config and structure that i use to test now. Can anyone please point me to a documentation how to properly debug my issue and where to look at?

https://github.com/jaklimoff/pyright_issue

![autoimport_screen](https://github.com/microsoft/pyright/assets/4048774/bf9ebc05-875b-442d-94fb-4ed9cd7b3449)

**Comment by erictraut** [comment, marked as answer]
The auto-import mechanism in pyright is relatively limited. It works off of its internal memory-based cache. If the type checker hasn't seen the file (either via an import or by opening the file), it will not have symbols for that file cached. 

Pyright is primarily focused on type checking. It has some limited language server features, but it's meant to be the foundation for more sophisticated language server features. Pylance is Microsoft's premium Python language server, and it includes many additional features. One of those features is an indexing mechanism that builds and maintains an index of symbols in all files in your project and all installed packages. This allows it to provide complete auto-import suggestions. If you want this functionality, I recommend switching to Pylance. It's licensed for use in VS Code.

If you want to stick with pyright, one workaround is to set the `python.analysis.diagnosticMode` setting from its default value "openFilesOnly" to "workspace". This will cause the language server to type check all of the files in your project. Unless your project is very large, all of the type information will fit in the in-memory cache, and all symbols in your project will be available for auto-import. If your project is very large, the cache may be deallocated periodically to make room for additional type information. In that case, you'll see inconsistent behaviors for auto-import.

**Reply by jaklimoff** [reply]
Thank you very much for quick and extensive answer I will try my luck with this approach with diagnostic mode. But how come it knows about symbols for libraries in my venv then if it is limited to open file?  Can i also just use `extraPaths = ["src"]` in config as an option?

**Reply by erictraut** [reply]
If it knows about symbols for a particular module in a particular library, that means it has encountered this module at some point during type checking — probably by following an existing import and its dependent imports. Pyright implements no persistent caching.
